Karl/karl.h

```diff
@@ -227,14 +316,68 @@ class GenericKarl final {
   }
 
   ~GenericKarl() {
+    destructing_ = true;
     storage_.ReadWriteTransaction([this](MutableFields<storage_t> fields) {
       KarlInfo self_info;
       self_info.up = false;
       fields.karl.Add(self_info);
     }).Wait();
+    if (state_update_thread_.joinable()) {
+      state_update_thread_.join();
+    }
+  }
+
+  size_t ActiveServicesCount() const {
+    std::lock_guard<std::mutex> lock(services_keepalive_cache_mutex_);
+    return services_keepalive_time_cache_.size();
   }
 
+  storage_t& InternalExposeStorage() { return storage_; }
+
  private:
+  void StateUpdateThread() {
+    while (!destructing_) {
+      const auto now = current::time::Now();
+      std::set<std::string> timeouted_codenames;
+      {
+        std::lock_guard<std::mutex> lock(services_keepalive_cache_mutex_);
+        for (auto it = services_keepalive_time_cache_.cbegin(); it != services_keepalive_time_cache_.cend();) {
+          if ((now - it->second) > up_threshold_) {
+            timeouted_codenames.insert(it->first);
+            services_keepalive_time_cache_.erase(it++);
+          } else {
+            ++it;
+          }
+        }
+      }
+      if (!timeouted_codenames.empty()) {
+        storage_.ReadWriteTransaction([&timeouted_codenames](MutableFields<storage_t> fields) -> void {
+          for (const auto& codename : timeouted_codenames) {
+            const auto& current_claire_info = fields.claires[codename];
+            ClaireInfo claire;
+            if (Exists(current_claire_info)) {
+              claire = Value(current_claire_info);
+            } else {
+              claire.codename = codename;
```

Comment:
:+1: perhaps mark this Claire as a zombie, even by the simplest way possible, by writing some `"zombie"` as its IP address?